src/ngMessages/messages.js

```diff
@@ -228,91 +291,217 @@ angular.module('ngMessages', [])
-        $scope.$watchCollection(watchAttr, function(values) {
-          cachedValues = values;
-          ctrl.renderMessages(values, multiple);
-        });
+             // dive deeper into the DOM and examine its children for any ngMessage
+             // comments that may be in an element that appears deeper in the list
+             if (prevNode.childNodes.length && parentLookup.indexOf(prevNode) == -1) {
+               parentLookup.push(prevNode);
+               prevNode = prevNode.childNodes[prevNode.childNodes.length - 1];
+             } else {
+               prevNode = prevNode.previousSibling || prevNode.parentNode;
+             }
+           }
+         }
 
-        var tpl = $attrs.ngMessagesInclude || $attrs.include;
-        if (tpl) {
-          $templateRequest(tpl)
-            .then(function processTemplate(html) {
-              var after, container = angular.element('<div/>').html(html);
-              angular.forEach(container.children(), function(elm) {
-               elm = angular.element(elm);
-               after ? after.after(elm)
-                     : element.prepend(elm); //start of the container
-               after = elm;
-               $compile(elm)($scope);
-              });
-              ctrl.renderMessages(cachedValues, multiple);
-            });
-        }
-      }
-    };
-  }])
+         function insertMessageNode(parent, comment, key) {
+           var messageNode = messages[key];
+           if (!ctrl.head) {
+             ctrl.head = messageNode;
+           } else {
+             var match = findPreviousMessage(parent, comment);
+             if (match) {
+               messageNode.next = match.next;
+               match.next = messageNode;
+             } else {
+               messageNode.next = ctrl.head.next;
+               ctrl.head = messageNode;
+             }
+           }
+         }
+
+         function removeMessageNode(parent, comment, key) {
+           var messageNode = messages[key];
+
+           var match = findPreviousMessage(parent, comment);
+           if (match) {
+             match.next = messageNode.next;
+           } else {
+             ctrl.head = messageNode.next;
+           }
+         }
+       }],
+
+       link: function(scope, element, attrs, ngMessagesCtrl) {
+         ngMessagesCtrl.render = function(collection) {
+           collection = collection || {};
+           var self = ngMessagesCtrl;
+           self._cachedCollection = collection;
+
+           // this is true if the attribute is empty or if the attribute value is truthy
+           var showAll = isAttrTruthy(scope, attrs.ngMessagesMultiple) ||
+                         isAttrTruthy(scope, attrs.multiple);
 
+           var matches = [];
+           var nameMatches = [];
+           var item = ngMessagesCtrl.head;
+           var messageExists = false;
+           while (item != null && !messageExists) {
```

Comment:
I tried that one. It doesn't work since if there are no messages present them item == undefined which !== null and then the loop never breaks.